### ISSUE TITLE
Establish and define our modified Gitflow process

### ISSUE THREAD

**Issue by mikemilano** [thinktandem/tandem#72]
> Note: *Github workflow* (aka Githubflow) and *Git workflow* (aka Gitflow) are different. *Github flow* is a vague derivative of *Gitflow*.

Objectives:
1. Discuss and adopt a more detailed Gitflow operating procedure
2. Document it in the 2.0 docs

Currently we define a [Github Workflow](https://guides.github.com/introduction/flow/) in the  process docs and lightly elaborate, but it would be helpful if we defined the Git workflow in detail as it applies to Tandem.

Why? Because the [Github Workflow](https://guides.github.com/introduction/flow/) doc is light on the concept it encapsulates, which is a Git workflow design "made popular by [Vincent Driessen at nvie](https://nvie.com/posts/a-successful-git-branching-model/)". (Source, [Atlassian](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow))

The common theme among most Gitflow examples is that the `master` branch is always the deployable branch, and the `develop` branch is typically where the latest integrations are merged into.

Strategy and naming conventions beyond this may change slightly by its practitioners based on a variety of project factors.

Note: Our workflow (in the old docs) is inconsistent with a Gitflow in that is there is no `develop` branch, and our docs instruct the developer to branch features directly off `master`. 

The concept of an integration branch (`develop`) is important to maintain changes towards a release while hotfixes can still be released to `master` without deploying any other work that is not ready to be released.

> Note: Different projects may merit different branching strategies. The one I'm referring to here is for agency website development work. Product development may or may not warrant different strategies based on how releases and collaboration happens.

I've seen us use a branch named `test` as a sort of integration branch, but it is still inconsistent with our existing docs which have us branch from `master`.

We should be using `develop` instead of `test` as it is an unofficial standard of sorts as seen in the references above. I'd also argue the name `test` has its own meaning and should not be used for a branch name unless it was a branch created to maintain an environment for automated tests.

Minimally, the following should be defined in this document.

- Branch naming (`master`, `develop`)
- Feature branches strategy (including naming conventions and acceptance process)
- Hotfix strategy
- Release strategy (i.e. tagging or branch names if applicable; This may even vary based on project params)
- Hosting provider specifics; For example, we may release differently on Platform than we do Pantheon based on Github integration constraints.

The challenge here is that I don't think we've ever gone into this much detail on process, so either someone needs to take a stab at this, or we need to organize an effort to collaborate.

Hey @pirog @reynoldsalec  @serundeputy @labboy0276 @dustinleblanc ! Tell me what you think!

**Comment by dustinleblanc** [comment]
The reasoning here sounds wise to me, so I agree. This *is* different than the way we normally work, but I don't have a lot of reasons to believe the way we currently work is any better other than being used to a master / feature-branch methodology.

The current method works very well if and only if we can deploy each feature to production as soon as it is merged.

It is worth noting that there are also [tools](https://github.com/nvie/gitflow) for making this flow easier, since it is more work.

**Comment by mikemilano** [comment]
@dustinleblanc, thanks for reviewing, and good point to analyze what's wrong with the current method. Defining what an integration branch does for us might be useful:

Benefits an integration branch gives us that we wouldn't have otherwise are:
- It is possible to internally review multiple features together before a release
  - i.e. check how your feature integrates against other un-released work being done in parallel
- Allows for a more formal (perhaps more responsible) and controlled releases
  - this could be releasing all work completed in a sprint at a planned date
  - or it could mean being able to stage a group of work for customer review, before releasing to master. of course staging happens on a staging environment, but we would have visibility into how the integration is coming along before we make it available on staging to clients.
- we have limited environments on platform, so activating environments for a bunch of feature branches isn't very feasible vs merging them into integration for review, before release

Also note that feature development can continue as it works now... the feature just being branched from `develop`.

**Comment by dustinleblanc** [comment]
Yep totally agree. I think this is a **smart** move. Its going to cause a bit of short term friction to change our 'muscle memory' but I think the benefits are pretty clear.

**Comment by labboy0276** [comment]
@mikemilano This all depends on what hosting provider in general.  The current workflow we have works just fine for p.sh.  We create branches from master and their PR's are where we do the QA we need.  There is no need for another branch like test or develop unless the project calls for it.

Outside of that, then I can see having a separate workflow

**Comment by mikemilano** [comment]
@labboy0276 thanks for reviewing, and good point. I agree, the current workflow does fine with small projects which have linear efforts and loose QA/release requirements. It breaks or becomes fragile when:

- Project requires multiple features to be tested/QA together before release
- Multiple developers are working on related features and could benefit from testing the integrated work
- PM or anyone internally needs to compare the latest unreleased WIP against production to see if a reported issue is resolved (before we formally stage)
- We want to pre-stage work before showing it to clients on a staging environment they have access to

The developer impact is negligible while the benefits allow us to support projects with a more formal CI and QA requirements.

The change to the process is this:
- features are branched from and merged into `develop` (hotfixes remain the same, from and to `master)
- releases are merges from `develop` to `master`, unless the project requires staging, in which case we merge `develop` to `staging` to `master`.

This simple change gives us much more control over what and when we release to production, along with greater visibility into the latest unreleased code across all features being worked on.

**Comment by mikemilano** [comment]
I had written these questions down before learning we were going to close this issue... but here are the questions I had for our current process does not support, and a small change to our branching strategy would resolve:

- How do we make sure multiple un-released features work together before releasing it to a staging environment where customers are looking and reporting issues against?

- How do we control the release of multiple features together?

- How does a developer or project lead confirm the PR they just merged integrates nicely on an environment after the merge, and before putting it in front of a customer? (again, considering there are multiple unreleased features)

- How does a developer working on a feature make sure their code works with the latest unreleased code?


If the answer to handle multiple features and controlled releases is something like: create a feature branch like `summer-launch`, and then branch new feature from and merge back into that, then it's essentially the same thing as the gitflow process that is proposed. It's just a standard way to maintain that level of control over pre-staging and releases.

**Comment by pirog** [comment]
So based on our conversation on Monday i am going to close this out and try to provide a task that is smaller in scope, addresses a single problem and is easier for us all to discuss. Basically the task is going to ask @mikemilano to digest this thread and propose a single change we can make to our current flow to make it better. 

That said i think it's probably important for you all to know where i stand on this so here goes. 

For starters i'd like to contextualize this response by saying i am imaging this for an "average tandem project" aka i am assuming platform.sh, QA environments for PRs, github integration. Said another way i think my response is true for _most_ of our projects. This implies a, perhaps philosophical, preference for the simplest process that covers most of our projects and then perhaps documented edge cases vs a "one size fits all" approach which i think would probably be overburdensome for most of our projects. For 100% clarity i am referring mostly to what is documented here
* https://docs.thinktandem.io/guides/contributing-code.html
* https://docs.thinktandem.io/guides/qaing-code.html
with the caveat that i think these guides definitely could use some work which i think this and subsequent issues should achieve. 

Now to get into the meat of things. So probably unsurprisingly given the above and since i set most of this flow up originally (not territorial i swear!) i agree more or less with @labboy0276. Specifically i think a lot of the benefits enumerated in favor of a `deploy` branch are already possible in our current model, generally dont happen a ton and do not come with what i would consider [unneeded complexity].(https://docs.thinktandem.io/manifesto/sins.html#marooning-useful-things) For example our current model does not require a separate hotfix flow, nor does it add an additional layer between PR environments and master which requires 1.) more administrative overhead 2.) potentially blocks production if PM mistakes are made. 3.) potentially slows down how often we deploy, which i think should be "whenever we can". 

That said and to be clear I 100% think there are projects where an additional layer eg a `deploy` branch is not only a good idea but necessary both because of the project itself and also the scale of people developing on it. However, with perhaps the exception of one project which i will not name because PUBLIC! i think this is likely and currently overkill for our other projects. I certainly do not think its caused any HUGE problems in the past. Also 100% admit that at some point in the future this may totally be our default flow.

So i think the below response is a single and easy thing that we can do (and already do in most cases) to get most of, if not all of the benefits listed 

Generally, although this is not explicitly documented (but should be), we setup a GitHub PR status that blocks a PR merge into master until said PR branch is up to date with master. So basically, a feature gets merged first and then automatically all other open features `require` master (which now contains the first feature) be merged before they can get deployed thus triggering another round of tests/qa/etc. In my opinion this effectively accomplishes the same thing as a `deploy/staging` branch except the "feature compatibility check" is decentralized to all other extant PR branches and does not have the disadvantage of a single pipe which may block things or complicate things like hotfixing. Additionally, enforcing master merges before other features can be merged means the QA environment for any other open feature that merges master functions as a place for both the dev and client to verify feature compatibility. Of course we should always be merging in master all the time, everywhere as a good dev practice but this also does enforce that it needs to be merged AT LEAST ONCE before shit hits the fan. Admittedly this does imply that all feature deployment happen sequentially but realistically (eg in practical terms) this happens in both models, you are just doing it in different places. Admittedly there is additional work to make sure all other open PRs are updated with master but 1.) easy to automate with bots if we want 2.) the UI makes it stupid simple 3.) if there are multiple things we need to merge at more or less exactly the same time (which doesnt really happen ever) we can just package them into a single PR beforehand eg at any point we can create a branch that is not centralized but whose purpose is to contain multiple features before they go to production. 

Anyway! those are my thoughts. @mikemilano im closing this issue out but if you feel compelled to respond please do so. That said, it might be BETTER and a GOOD EXERCISE for me to spin up this new ticket and for you to scope it out, digest all that is here in this ticket and then make a judgement call of how best we can move forward given the concerns and feedback. 

I'll get on that next ticket right away!

**Comment by pirog** [comment]
ok @mikemilano ive queued up #90
